File: Tests/CSharp/Analyzers/FaultEffectAnalyzer.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Analyzers

open System
open System.Linq
open NUnit.Framework
open SafetySharp.Modeling
open SafetySharp.CSharp.Analyzers
open SafetySharp.CSharp.Roslyn.Syntax
open SafetySharp.CSharp.Roslyn.Symbols

[<TestFixture>]
module ``Fault effects`` =
    let getDiagnostic csharpCode = 
        let csharpCode = 
            sprintf "
                class X : Y { 
                    public void Void2Void() { }
                    public int Int2Int(int x) { return x; }
                    public void Overloaded(int x) { }
                    public void Overloaded() { }
                    public void Overloaded(bool b) { }
                    public void Ref(ref int x) { }
                    public void Out(out int x) { x = 1; }
                    public void Optional(int x = 3) { }
                    public int GetProp => 1;
                    public int SetProp { set { } }
                    public int Prop { get; set; }
                    public int get_GetMethodProp() => 1;
                    public void set_SetMethodProp(int value) { }
                    [Transient] class F : Fault { %s }
                }
                class Y : Component { 
                    public void Inherited() { }
                    private void Inaccessible() { }
                    public int InheritedProp => 1;
                    private int InaccessibleProp => 1;
                }
            " csharpCode
        TestCompilation.GetDiagnostic (FaultEffectAnalyzer ()) csharpCode

    let getLocation location = location + 50
    let line = 16

    let unknownMethodPort location memberName effect =
        errorDiagnostic DiagnosticIdentifier.FaultEffectUnknownMethodPort (line, getLocation location) (line, getLocation location + (String.length memberName))
            "Invalid fault method '%s': Affected component of type 'X' does not declare a method named '%s'." effect memberName

    let unknownPropertyPort location memberName effect =
        errorDiagnostic DiagnosticIdentifier.FaultEffectUnknownPropertyPort (line, getLocation location) (line, getLocation location + (String.length memberName))
            "Invalid fault property '%s': Affected component of type 'X' does not declare a property named '%s' with the corresponding accessors." effect memberName

    let signatureIncompatible location memberName effect portName =
        errorDiagnostic DiagnosticIdentifier.FaultEffectSignatureIncompatible (line, getLocation location) (line, getLocation location + (String.length memberName))
            "Fault effect '%s' is not signature compatible with '%s' or any of its overloads." effect portName

    let optionalParameter location effect parameterName =
        errorDiagnostic DiagnosticIdentifier.FaultEffectOptionalParameter (line, getLocation location) (line, getLocation location + (String.length parameterName))
            "Invalid optional parameter '%s' declared by fault effect '%s'." parameterName effect

    let baseMember location memberName effect baseClass =
        errorDiagnostic DiagnosticIdentifier.FaultEffectBaseMember (line, getLocation location) (line, getLocation location + (String.length memberName))
            "Fault effect '%s' cannot affect inherited member '%s'." effect baseClass

    let propertyType location memberName effect propertyType =
        errorDiagnostic DiagnosticIdentifier.FaultEffectIncompatibleType (line, getLocation location) (line, getLocation location + (String.length memberName))
            "Fault effect '%s' must be a property of type '%s'." effect propertyType

    [<Test>]
    let ``fault class without effects is valid`` () =
        getDiagnostic "" =? None

    [<Test>]
    let ``non-nested fault class is valid`` () =
        TestCompilation.GetDiagnostic (FaultEffectAnalyzer ()) "[Transient] class F : Fault { public void M() {} }" =? None

    [<Test>]
    let ``fault class nested in non-component class is valid`` () =
        TestCompilation.GetDiagnostic (FaultEffectAnalyzer ()) "class X { [Transient] class F : Fault { public void M() {} }}" =? None

    [<Test>]
    let ``fault effect for unknown port is invalid`` () =
        getDiagnostic "public void Q() {}" =? unknownMethodPort 12 "Q" "X.F.Q()" 

    [<Test>]
    let ``fault effect for signature compatible port is valid`` () =
        getDiagnostic "public void Void2Void() { }" =? None
        getDiagnostic "public int Int2Int(int a) { return a; }" =? None
        getDiagnostic "public void Overloaded(int b) { }" =? None
        getDiagnostic "public void Overloaded() { }" =? None
        getDiagnostic "public void Overloaded(bool d) { }" =? None
        getDiagnostic "public void Ref(ref int q) { }" =? None
        getDiagnostic "public void Out(out int q) { q = 1; }" =? None

    [<Test>]
    let ``fault effect for property port getter or setter with same type is valid`` () =
        getDiagnostic "public int GetProp => 1;" =? None
        getDiagnostic "public int SetProp { set { } }" =? None
        getDiagnostic "public int Prop { get; set; }" =? None

    [<Test>]
    let ``fault effect for signature incompatible port is valid`` () =
        getDiagnostic "public void Void2Void(int i) { }" =? signatureIncompatible 12 "Void2Void" "X.F.Void2Void(int)" "X.Void2Void()"
        getDiagnostic "public int Void2Void() { return 0; }" =? signatureIncompatible 11 "Void2Void" "X.F.Void2Void()" "X.Void2Void()"
        getDiagnostic "public int Int2Int(bool a) { return 1; }" =? signatureIncompatible 11 "Int2Int" "X.F.Int2Int(bool)" "X.Int2Int(int)"
        getDiagnostic "public bool Int2Int(int a) { return false; }" =? signatureIncompatible 12 "Int2Int" "X.F.Int2Int(int)" "X.Int2Int(int)"
        getDiagnostic "public void Int2Int() { }" =? signatureIncompatible 12 "Int2Int" "X.F.Int2Int()" "X.Int2Int(int)"
        getDiagnostic "public bool Overloaded() { return false; }" =? signatureIncompatible 12 "Overloaded" "X.F.Overloaded()" "X.Overloaded(int)"
        getDiagnostic "public void Ref(out int q) { q = 0; }" =? signatureIncompatible 12 "Ref" "X.F.Ref(out int)" "X.Ref(ref int)"
        getDiagnostic "public void Out(ref int q) { q = 1; }" =? signatureIncompatible 12 "Out" "X.F.Out(ref int)" "X.Out(out int)"
        getDiagnostic "public void Ref(int q) { q = 0; }" =? signatureIncompatible 12 "Ref" "X.F.Ref(int)" "X.Ref(ref int)"
        getDiagnostic "public void Out(int q) { q = 1; }" =? signatureIncompatible 12 "Out" "X.F.Out(int)" "X.Out(out int)"

    [<Test>]
    let ``fault effect for property with different type is invalid`` () =
        getDiagnostic "public bool GetProp => false;" =? propertyType 12 "GetProp" "X.F.GetProp" "int"
        getDiagnostic "public bool SetProp { set { } }" =? propertyType 12 "SetProp" "X.F.SetProp" "int"
        getDiagnostic "public bool Prop { get; set; }" =? propertyType 12 "Prop" "X.F.Prop" "int"

    [<Test>]
    let ``fault effect overriding port with optional parameter is valid if it has no optional parameter`` () =
        getDiagnostic "public void Optional(int g) { }" =? None

    [<Test>]
    let ``fault effect with optional parameter is invalid`` () =
        getDiagnostic "public void Optional(int g = 4) { }" =? optionalParameter 25 "X.F.Optional(int)" "g"
        getDiagnostic "public int Int2Int(int x = 2) { return x; }" =? optionalParameter 23 "X.F.Int2Int(int)" "x"

    [<Test>]
    let ``fault effect with generic parameter of generic component is valid`` () =
        TestCompilation.GetDiagnostic (FaultEffectAnalyzer ()) "class X<T> : Component { public T M(T t) { return t; } [Transient] class F : Fault { public T M(T t) { return t; } }}" =? None

    [<Test>]
    let ``fault effect overriding accessible base member is invalid`` () =
        getDiagnostic "public void Inherited() { }" =? baseMember 12 "Inherited" "X.F.Inherited()" "Y.Inherited()"
        getDiagnostic "public int InheritedProp => 1;" =? baseMember 11 "InheritedProp" "X.F.InheritedProp" "Y.InheritedProp"

    [<Test>]
    let ``fault effect overriding inaccessible base member is invalid`` () =
        getDiagnostic "public void Inaccessible() { }" =? unknownMethodPort 12 "Inaccessible" "X.F.Inaccessible()"
        getDiagnostic "public int InaccessibleProp => 1;" =? unknownPropertyPort 11 "InaccessibleProp" "X.F.InaccessibleProp"

    [<Test>]
    let ``fault effect property that overrides method is invalid`` () =
        getDiagnostic "public int GetMethodProp => 1;" =? unknownPropertyPort 11 "GetMethodProp" "X.F.GetMethodProp"
        getDiagnostic "public int SetMethodProp { set {} }" =? unknownPropertyPort 11 "SetMethodProp" "X.F.SetMethodProp"

    [<Test>]
    let ``fault effect method that overrides property is invalid`` () =
        getDiagnostic "public int GetProp() => 1;" =? unknownMethodPort 11 "GetProp" "X.F.GetProp()" 
        getDiagnostic "public void SetProp(int value) {}" =? unknownMethodPort 12 "SetProp" "X.F.SetProp(int)" 

    [<Test>]
    let ``overriding get-only property with get-set fault effect is invalid`` () =
        getDiagnostic "public int GetProp { get; set; }" =? unknownPropertyPort 11 "GetProp" "X.F.GetProp"

    [<Test>]
    let ``overriding set-only property with get-set fault effect is invalid`` () =
        getDiagnostic "public int SetProp { get; set; }" =? unknownPropertyPort 11 "SetProp" "X.F.SetProp"

    [<Test>]
    let ``overriding get-set property with get-only fault effect is valid`` () =
        getDiagnostic "public int Prop => 1;" =? None

    [<Test>]
    let ``overriding get-set property with set-only fault effect`` () =
        getDiagnostic "public int Prop { set {} }" =? None

File: Source/Compiler/Analyzers/FaultEffectAnalyzer.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.Compiler.Analyzers
{
	using System;
	using System.Linq;
	using JetBrains.Annotations;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.Diagnostics;
	using Roslyn;
	using Roslyn.Symbols;

	/// <summary>
	///     Ensures that fault effects are declared correctly.
	/// </summary>
	[DiagnosticAnalyzer(LanguageNames.CSharp), UsedImplicitly]
	public sealed class FaultEffectAnalyzer : Analyzer
	{
		/// <summary>
		///     Indicates that the fault effect overrides an unknown method port.
		/// </summary>
		private static readonly DiagnosticInfo _unknownMethodPort = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectUnknownMethod,
			"Fault effects must override existing methods of the affected component.",
			"Invalid fault effect '{0}': Affected component of type '{1}' does not declare a method named '{2}'.");

		/// <summary>
		///     Indicates that the fault effect overrides an unknown property port.
		/// </summary>
		private static readonly DiagnosticInfo _unknownPropertyPort = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectUnknownProperty,
			"Fault effects must override existing property of the affected component.",
			"Invalid fault effect '{0}': Affected component of type '{1}' does not declare a property named '{2}' with the corresponding accessors.");

		/// <summary>
		///     Indicates that the member the fault effect overrides is ambiguous.
		/// </summary>
		private static readonly DiagnosticInfo _ambiguousMember = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectAmbiguousMember,
			"The fault effect's affected member is ambiguous.",
			"Invalid fault effect '{0}': Affected member is ambiguous; could be {1}.");

		/// <summary>
		///     Indicates that the fault effect overrides an unknown port.
		/// </summary>
		private static readonly DiagnosticInfo _signatureIncompatible = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectSignatureIncompatible,
			"Fault effects must be signature compatible with the overriden port of the affected component.",
			"Fault effect '{0}' is not signature compatible with '{1}' or any of its overloads.");

		/// <summary>
		///     Indicates that the fault effect overrides a property port of a different type.
		/// </summary>
		private static readonly DiagnosticInfo _typeIncompatible = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectIncompatibleType,
			"Fault effect properties must be declared with the type of the overriden property port of the affected component.",
			"Fault effect '{0}' must be a property of type '{1}'.");

		/// <summary>
		///     Indicates that the fault effect overrides a member of a base class.
		/// </summary>
		private static readonly DiagnosticInfo _baseMember = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectBaseMember,
			"Fault effects can only override members declared by the affected component.",
			"Fault effect '{0}' cannot affect inherited member '{1}'.");

		/// <summary>
		///     Indicates that the fault effect declares an optional parameter.
		/// </summary>
		private static readonly DiagnosticInfo _optionalParameter = DiagnosticInfo.Error(
			DiagnosticIdentifier.FaultEffectOptionalParameter,
			"Fault effects cannot declare optional parameters.",
			"Invalid optional parameter '{0}' declared by fault effect '{1}'.");

		/// <summary>
		///     Indicates that the fault declares no effects.
		/// </summary>
		private static readonly DiagnosticInfo _noEffects = DiagnosticInfo.Warning(
			DiagnosticIdentifier.FaultWithoutEffects,
			"The fault does not declare any fault effects.",
			"Fault '{0}' does not declare any fault effects.");

		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		public FaultEffectAnalyzer()
			: base(
				_unknownMethodPort, _unknownPropertyPort, _signatureIncompatible, _baseMember, _optionalParameter,
				_typeIncompatible, _noEffects, _ambiguousMember)
		{
		}

		/// <summary>
		///     Called once at session start to register actions in the analysis context.
		/// </summary>
		/// <param name="context" />
		public override void Initialize(AnalysisContext context)
		{
			context.RegisterSymbolAction(Analyze, SymbolKind.NamedType);
		}

		/// <summary>
		///     Performs the analysis.
		/// </summary>
		/// <param name="context">The context in which the analysis should be performed.</param>
		private static void Analyze(SymbolAnalysisContext context)
		{
			var faultSymbol = context.Symbol as ITypeSymbol;
			if (faultSymbol == null || !faultSymbol.IsDerivedFromFault(context.Compilation))
				return;

			var componentSymbol = faultSymbol.ContainingType;
			if (componentSymbol == null || !componentSymbol.IsDerivedFromComponent(context.Compilation))
				return;

			var faultProperties = faultSymbol.GetMembers().OfType<IPropertySymbol>().ToArray();
			var faultMethods = faultSymbol
				.GetMembers()
				.OfType<IMethodSymbol>()
				.Where(candidate => candidate.MethodKind == MethodKind.Ordinary && !candidate.IsUpdateMethod(context.Compilation))
				.ToArray();

			if (faultMethods.Length == 0 && faultProperties.Length == 0)
				_noEffects.Emit(context, faultSymbol, faultSymbol.ToDisplayString());

			foreach (var method in faultMethods)
			{
				var optionalParameter = method.Parameters.FirstOrDefault(parameter => parameter.IsOptional);
				if (optionalParameter != null)
					_optionalParameter.Emit(context, optionalParameter, optionalParameter.Name, method.ToDisplayString());

				var candidateMethods = componentSymbol
					.GetMembers()
					.OfType<IMethodSymbol>()
					.Where(candidate => candidate.MethodKind == MethodKind.Ordinary || candidate.MethodKind == MethodKind.ExplicitInterfaceImplementation)
					.Where(candidate => candidate.MethodKind == MethodKind.ExplicitInterfaceImplementation
						? candidate.ExplicitInterfaceImplementations[0].Name == method.Name
						: candidate.Name == method.Name)
					.ToArray();

				if (candidateMethods.Length == 0)
				{
					// There is no port with a matching name; there might be two reasons:
					// - The affected component doesn't declare a port with the given name
					// - The component inherited, but did not redeclare or override a port with the given name
					// Even though a generic "port not found" message is enough for both cases, we improve the
					// diagnosis by giving the user more context

					var baseType = componentSymbol.BaseType;
					IMethodSymbol[] baseCandidateMethods = null;

					while (baseType != null && (baseCandidateMethods == null || baseCandidateMethods.Length > 1))
					{
						baseCandidateMethods = baseType
							.GetMembers(method.Name)
							.OfType<IMethodSymbol>()
							.Where(candidate => componentSymbol.CanAccessBaseMember(candidate))
							.ToArray();

						baseType = baseType.BaseType;
					}

					if (baseCandidateMethods != null && baseCandidateMethods.Length > 0)
						_baseMember.Emit(context, method, method.ToDisplayString(), baseCandidateMethods[0].ToDisplayString());
					else
						_unknownMethodPort.Emit(context, method, method.ToDisplayString(), componentSymbol.ToDisplayString(), method.Name);
				}
				else
				{
					var signatureCompatibleMethods = candidateMethods.Where(candidate => candidate.IsSignatureCompatibleTo(method)).ToArray();

					if (signatureCompatibleMethods.Length > 1)
					{
						_ambiguousMember.Emit(context, method, method.ToDisplayString(),
							String.Join(", ", candidateMethods.Select(m => String.Format("'{0}'", m.ToDisplayString()))));
					}
					else if (signatureCompatibleMethods.Length == 0)
						_signatureIncompatible.Emit(context, method, method.ToDisplayString(), candidateMethods[0].ToDisplayString());
				}
			}

			foreach (var property in faultProperties)
			{
				var candidateProperties = componentSymbol
					.GetMembers()
					.OfType<IPropertySymbol>()
					.Where(candidate => candidate.ExplicitInterfaceImplementations.Length != 0
						? candidate.ExplicitInterfaceImplementations[0].Name == property.Name
						: candidate.Name == property.Name)
					.ToArray();

				if (candidateProperties.Length == 0)
				{
					// There is no port with a matching name; there might be two reasons:
					// - The affected component doesn't declare a port with the given name
					// - The component inherited, but did not redeclare or override a port with the given name
					// Even though a generic "port not found" message is enough for both cases, we improve the
					// diagnosis by giving the user more context

					var baseType = componentSymbol.BaseType;
					IPropertySymbol baseProperty = null;

					while (baseType != null && baseProperty == null)
					{
						baseProperty = baseType
							.GetMembers(property.Name)
							.OfType<IPropertySymbol>()
							.FirstOrDefault(candidate => componentSymbol.CanAccessBaseMember(candidate));

						baseType = baseType.BaseType;
					}

					if (baseProperty != null)
						_baseMember.Emit(context, property, property.ToDisplayString(), baseProperty.ToDisplayString());
					else
						_unknownPropertyPort.Emit(context, property, property.ToDisplayString(), componentSymbol.ToDisplayString(), property.Name);
				}
				else if (candidateProperties.Length > 1)
				{
					_ambiguousMember.Emit(context, property, property.ToDisplayString(),
						String.Join(", ", candidateProperties.Select(m => String.Format("'{0}'", m.ToDisplayString()))));
				}
				else
				{
					var candidateProperty = candidateProperties[0];

					if (!property.Type.Equals(candidateProperty.Type))
						_typeIncompatible.Emit(context, property, property.ToDisplayString(), candidateProperty.Type.ToDisplayString());

					var faultHasGetter = property.GetMethod != null;
					var faultHasSetter = property.SetMethod != null;
					var componentHasGetter = candidateProperty.GetMethod != null;
					var componentHasSetter = candidateProperty.SetMethod != null;

					var invalidGetter = !componentHasGetter && faultHasGetter;
					var invalidSetter = !componentHasSetter && faultHasSetter;

					if (invalidGetter || invalidSetter)
						_unknownPropertyPort.Emit(context, property, property.ToDisplayString(), componentSymbol.ToDisplayString(), property.Name);
				}
			}
		}
	}
}